### ISSUE TITLE
How to efficiently bulk index documents

### ISSUE THREAD

**Issue by MickVermeulen** [sabricot/django-elasticsearch-dsl#205]
How can you efficiently index a large number of Django models in code? Currently I'm using `ItemDocument().update(Item.objects.all())` but for ~1.000.000 models this takes over an hour to index. Is there a more efficient way to tackle this problem?

**Comment by josh-stableprice** [comment]
I don't know what your use case is, but one way is to not do it all at once, try to find a way in which you only alter the instances as/when changes happen

**Comment by MickVermeulen** [comment]
We get a feed with about 500.000 entries which we process in batches of 15.000 at a time. However, as we have to do a cleanup after processing the 500.000 entries we have to do the indexing after the cleanup. We can probably change that pipeline but we were wondering if there were any ways of bulk updating at a more efficient pace. Running `search_index --rebuild`, also takes a while but not nearly as long as caling `ItemDocument().update(Item.objects.all())`. But as far as you know there are not more efficient ways to bulk indexing besides calling `Document.update()`?

**Comment by alexgarel** [comment]
On my side, we index lot of complicated documents.

To be more efficient this is what we have done:

- use prefetch_related / select_related on the queryset to have all the data we need fast
- if we have some complex computation for one object, used for different properties, cache it in the Document class
- we have a management command that launch multiple process, and chunks the work among them
[^2]

And some advanced stuff:
- rewrite the _get_actions method to use our own pager  [^1]
- this pager optimize the query, by first retrieving pks and then batching on them (in our case it is faster because the query is very complex and postgres struggles to paginate it easily)
- also this pager, before releasing the page back to the indexer,  batch precompute some properties, that are faster computed as batch, and we put them in an special attribute of the Document instance, with precomputed properties in a dict indexed by each object pk


[^1]: I should have written a PR long ago to have the pager instantiation as a method on Document, so that it's easier to overload it, but did not did it yet :-/
[^2]: I could try to contribute something like this, but we have to make some assumption to group easily

**Comment by safwanrahman** [comment]
There is a fix in  #154 for speedup, but its not perfect. I think you can slice the queryset and run in different terminal. This is what I have done in my workplace.

So open about ten terminal, and do something like in one terminal and increase the number. like `1-10000`, `10001-20001` etc.
```python
objects = Item.objects.all().order_by('id')[1:10000]
ItemDocument().update(objects)
```
